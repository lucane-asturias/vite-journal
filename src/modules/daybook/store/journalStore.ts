import { defineStore } from 'pinia'
import { journalApi } from '@/apis'

export type Entry = { id?: string, date: number, text: string, picture?: undefined | string }
type EntriesResponse = { date: number, text: string, picture?: null | string }

interface JournalState {
  entries: Entry[]
  isLoading: boolean
}

export const useJournalStore = defineStore('journalStore', {
  state: () => ({
    entries: [],
    isLoading: true
  }) as JournalState,
  actions: {
    async loadEntries() {
      const { data } = await journalApi.get<EntriesResponse>('/entries.json')

      if (!data) {
        this.entries = []
        this.isLoading = false
        return
      }

      const entries = []

      for (let id of Object.keys(data)) {
        entries.push({
          id,
          ...data[id]
        })
      }

      this.entries = [ ...this.entries, ...entries ]
      this.isLoading = false
    },
    async createEntry({ text, date, picture }: Entry) {
      const dataToSave = { text, date, picture }
      // const teste = '' // <<
      const { data } = await journalApi.post(`/entries.json`, dataToSave)

      Object.assign(dataToSave, { id: data.name })  // data = { "name": "-MxlYBvjVoYTMaMJG7mt" }
      
      this.entries = [ dataToSave, ...this.entries ]

      return data.name
    },
    async updateEntry({ id, text, date, picture }: Entry) {
      const dataToSave = { date, picture, text }

      const response = await journalApi.put(`/entries/${ id }.json`, dataToSave)
      Object.assign(dataToSave, { id }) // sending id for vitest purposes

      const idx = this.entries.map(e => e.id).indexOf(id)
      this.entries[idx] = dataToSave
    },
    async destroyEntry(payloadEntryId: string) {
      await journalApi.delete(`/entries/${ payloadEntryId }.json`)
      this.entries = this.entries.filter(entry => entry.id !== payloadEntryId)
    }
  },
  getters: {
    getEntriesByTerm: (state) => {
      return (term: string): object => {
        if (term?.length === 0) return state.entries

        return state.entries.filter(
          entry => entry.text.toLowerCase().includes(term.toLowerCase())
        )
      }
    }, 
    getEntriesById: (state) => {
      return (entryId: string): object => {
        const entryObj = state.entries.find(entry => entry.id === entryId)
        if (!entryObj) return

        return { ...entryObj }
      } 
    } 
  }
})