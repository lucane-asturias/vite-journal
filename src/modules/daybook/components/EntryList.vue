<template>
  <div class="entry-list-container">
    <div class="p-2">
      <input 
        type="text"
        class="form-control"
        placeholder="Buscar entrada"
        v-model="data.term"
      />
    </div>

    <div class="mt-2 d-flex flex-column">
      <button class="btn btn-primary mx-3"
        @click="$router.push({ name: 'entry', params: { id: 'new' } })">
        <i class="fa fa-plus-circle">
          Nova entrada
        </i>
      </button>
    </div>

    <div class="entry-scrollarea">
      <Entry
        v-for="entry in entriesByTerm"
        :key="entry.id"
        :entry="entry"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
  import { computed, defineAsyncComponent, reactive } from 'vue'
  import { Entry as EntryType } from '../store/journalStore'
  import { useJournal } from '../composables/useJournal'

  const Entry = defineAsyncComponent(() => import('../components/Entry.vue'))

  const { getEntriesByTerm } = useJournal() 

  const data = reactive({ term: '' })

  const entriesByTerm = computed<EntryType>(() => getEntriesByTerm(data.term)) 

</script>

<style lang="scss" scoped>
.entry-list-container{
  border-right: 1px solid #2c3e50;
  height: calc( 100vh - 56px );
}

.entry-scrollarea{
  height: calc( 100vh - 110px );
  overflow: scroll;
}
</style>
