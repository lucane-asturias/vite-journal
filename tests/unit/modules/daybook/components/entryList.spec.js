import { beforeEach, describe, expect, test, vi } from "vitest";
import { shallowMount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { createTestingPinia } from '@pinia/testing'

import { useJournalStore } from '@/modules/daybook/store/journalStore'
import EntryList from '@/modules/daybook/components/EntryList.vue'

import { journalMockState } from '../../../mock-data/test-journal-state'


describe('Porvas no componente EntryList', () => {
  const mockRouter = { push: vi.fn() }

  let wrapper
  // SETUP - run prior to each unit test
  beforeEach(() => {
    // creates a fresh pinia and make it active so it's automatically picked
    // up by any useStore() call without having to pass it to it:
    // `useStore(pinia)`
    setActivePinia(createPinia())

    wrapper = shallowMount(EntryList, {
      global: {
        mocks: { $router: mockRouter },
        plugins: [
          createTestingPinia({ 
            createSpy: vi.fn,
            initialState: {
              journalStore: { ...journalMockState }
            }
          })
        ]
      }
    })

    vi.clearAllMocks()
  })

  test('deve chamar getEntriesByTerm sem termo e mostrar 2 entradas', () => {
    expect( wrapper.findAll('entry-stub').length ).toBe(2)
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('deve chamar getEntriesByTerm e filtrar as entradas', async () => {
    const input = wrapper.find('input')
    await input.setValue('segunda')

    expect(wrapper.findAll('entry-stub').length).toBe(1)
  })
  
  test('o botÃ£o de nova entrada deve redirecionar a /new', () => {
    wrapper.find('button').trigger('click')

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'entry', params: { id: 'new' } })
  })

})
