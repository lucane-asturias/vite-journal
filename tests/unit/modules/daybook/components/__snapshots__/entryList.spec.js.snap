// Vitest Snapshot v1

exports[`Porvas no componente EntryList > deve chamar getEntriesByTerm sem termo e mostrar 2 entradas 1`] = `
"<div class=\\"entry-list-container\\" data-v-d2f66aca=\\"\\">
  <div class=\\"p-2\\" data-v-d2f66aca=\\"\\"><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Buscar entrada\\" data-v-d2f66aca=\\"\\"></div>
  <div class=\\"mt-2 d-flex flex-column\\" data-v-d2f66aca=\\"\\"><button class=\\"btn btn-primary mx-3\\" data-v-d2f66aca=\\"\\"><i class=\\"fa fa-plus-circle\\" data-v-d2f66aca=\\"\\"> Nova entrada </i></button></div>
  <div class=\\"entry-scrollarea\\" data-v-d2f66aca=\\"\\">
    <entry-stub entry=\\"[object Object]\\" data-v-d2f66aca=\\"\\"></entry-stub>
    <entry-stub entry=\\"[object Object]\\" data-v-d2f66aca=\\"\\"></entry-stub>
  </div>
</div>"
`;
